blog post: Julia 1.6 Highlights (#1135)
JuliaLang/www.julialang.org: blog/2021/02/julia-1.6-highlights.md

are needed for the future health of the ecosystem made it into the release. The
final decision about whether Julia 1.6 will become the new LTS will not be made
until around the time that the 1.7 release enters stabilization, at which point
we'll have a much better sense of how good of a release it was — like baked
goods, some releases just come out better than others and you can't always tell
until after they are baked and eaten. The full list of changes can be found in
the [NEWS file](https://github.com/JuliaLang/julia/blob/release-1.6/NEWS.md),
but here we'll give a more in depth overview of some of the release highlights.

\toc

## Parallel precompilation

_Ian Butterworth_

Executing all of the statements in a module often involves compiling a large
amount of code, so Julia creates precompiled caches of the module to reduce this
time. In 1.6, this package precompilation is faster and happens before you leave
`pkg>` mode. Until now, precompilation took place solely as a single-processed
sequence, precompiling dependencies one-by-one when needed during the linear
code loading process when a package is `using`/`import`-ed for the first time.

> The olden days of <= 1.5:

Reviewer: ```suggestion

The olden days of <= 1.5:
```

> ```jl

Reviewer: ```suggestion
```julia-repl
```

(v1.5) pkg> add DifferentialEquations
...
julia> @time using DifferentialEquations
[ Info: Precompiling DifferentialEquations [0c46a032-eb83-5123-abaf-570d42b7fbaa]

>   474.288251 seconds ...

Reviewer: ```suggestion
  474.288251 seconds …
```

```
In 1.6, `pkg>` mode gains a heavily parallelized precompile operation that is
auto-invoked after package actions, to keep the active environment ready to
load.

> ```jl

Reviewer: ```suggestion
```julia-repl
```

mechanism took 1m12s to precompile while showing progress through the
dependencies, then the first time the package is loaded it happens at full
speed. The new parallel precompilation process takes a depth-first approach of
working through the dependency tree in the manifest, first precompiling the
packages with no dependencies and working upward to the packages listed in the
environment’s `Project.toml`, allowing multiple packages to be precompiled at
the same time. The operation is multi-processed, as opposed to multi-threaded,
so is not limited by Julia’s thread count. By default, Julia will spawn a
maximum CPU-balanced load of package precompile jobs at once based on the number
of CPU cores. Errors during precompilation will only throw for packages listed
in the Project to allow for dependencies that may be listed in manifests but not
loaded, and the auto precompilation process will remember if a package has
errored within the given environment and will not retry until it changes.
Auto-precompilation can be gracefully interrupted with a `ctrl-c` and disabled
by setting the environment variable `JULIA_PKG_PRECOMPILE_AUTO=0`.

> <script id="asciicast-381203" src="https://asciinema.org/a/381203.js" async></script>

Reviewer: ```suggestion
~~~
<script id="asciicast-381203" src="https://asciinema.org/a/381203.js" async></script>
~~~
```

For packages that are being developed, given that their code will be changed by other mechanisms than Pkg, this new workflow won’t automatically avoid encountering the standard code-load time precompilation. However non-dev-ed dependencies of those packages will be kept ready to load, so top-level precompilation at load time should remain lower for dev-ed packages.

> [Insert table of example precomp speed improvements?]

Reviewer: ```suggestion
```
Perhaps further examples aren't needed

## Compile time percentage

_Ian Butterworth_

A small change that should help understanding of one of Julia’s quirks for
newcomers is that the timing macro `@time` and its verbose friend `@timev` now
report if any of the reported time has been spent on compilation.[^1]

> ```jl

Reviewer: ```suggestion
```julia-repl
```

julia> x = rand(10,10);

julia> @time x * x;
  0.540600 seconds (2.35 M allocations: 126.526 MiB, 4.43% gc time, 99.94% compilation time)

julia> @time x * x;
  0.000010 seconds (1 allocation: 896 bytes)
```
Given Julia’s Just In Time (JIT) / Just Ahead Of Time (JAOT) compilation, the
first time code is run the compilation overhead is often substantial, with big
speed improvements seen in subsequent calls. This change highlights that
behavior, serving as both a reminder and a tool for rooting out unwanted
compilation effort i.e. over-specialized code

> [^1]: Note that the time reported by @time does not include time spent in
> julia’s code interpreter, which may be significant.

Reviewer: ```suggestion
[^1]: Note that the time reported by @time does not include time spent in julia’s code interpreter, which may be significant.
```

are needed for the future health of the ecosystem made it into the release. The
final decision about whether Julia 1.6 will become the new LTS will not be made
until around the time that the 1.7 release enters stabilization, at which point
we'll have a much better sense of how good of a release it was — like baked
goods, some releases just come out better than others and you can't always tell
until after they are baked and eaten. The full list of changes can be found in
the [NEWS file](https://github.com/JuliaLang/julia/blob/release-1.6/NEWS.md),
but here we'll give a more in depth overview of some of the release highlights.

\toc

## Parallel precompilation

_Ian Butterworth_

Executing all of the statements in a module often involves compiling a large
amount of code, so Julia creates precompiled caches of the module to reduce this
time. In 1.6, this package precompilation is faster and happens before you leave
`pkg>` mode. Until now, precompilation took place solely as a single-processed
sequence, precompiling dependencies one-by-one when needed during the linear
code loading process when a package is `using`/`import`-ed for the first time.

The olden days of <= 1.5:

> ```julia-repl

Reviewer: @tlienart any idea why we're not getting the nice pkg repl highlighting with this code tag? https://julialang.netlify.app/previews/pr1135/blog/2021/02/julia-1.6-highlights/#parallel_precompilation

Reviewer: Maybe Franklin needs updating https://github.com/JuliaLang/www.julialang.org/pull/1136

Reviewer: Resolved

mechanism took 1m12s to precompile while showing progress through the
dependencies, then the first time the package is loaded it happens at full
speed. The new parallel precompilation process takes a depth-first approach of
working through the dependency tree in the manifest, first precompiling the
packages with no dependencies and working upward to the packages listed in the
environment’s `Project.toml`, allowing multiple packages to be precompiled at
the same time. The operation is multi-processed, as opposed to multi-threaded,
so is not limited by Julia’s thread count. By default, Julia will spawn a
maximum CPU-balanced load of package precompile jobs at once based on the number
of CPU cores. Errors during precompilation will only throw for packages listed
in the Project to allow for dependencies that may be listed in manifests but not
loaded, and the auto precompilation process will remember if a package has
errored within the given environment and will not retry until it changes.
Auto-precompilation can be gracefully interrupted with a `ctrl-c` and disabled
by setting the environment variable `JULIA_PKG_PRECOMPILE_AUTO=0`.

~~~

> <script id="asciicast-381203" src="https://asciinema.org/a/381203.js" async></script>

Reviewer: Fit on page better & smaller (I don't have commit rights, so would someone mind accepting this)
```suggestion
<script id="asciicast-381203" src="https://asciinema.org/a/381203.js" data-rows="25" data-cols="104" async></script>
```

there were many places where Julia and its standard libraries were written in a
way that defeated Julia’s type-inference. Because the compiler sometimes had to
invalidate code just because a new method _might_ apply, any uncertainty about
types magnifies the risk and frequency of invalidation. In older versions of
Julia, the combination of these effects made invalidation widespread: just
loading certain packages led to invalidation of up to 10% of Julia’s precompiled
code. The delay for recompilation could sometimes make interactive sessions feel
sluggish. When invalidation occurred in Julia’s package-loading code, it also
delayed loading of the next package, contributing to long waits for using
SomePkg when SomePkg depends on other packages.

In 1.6, the scheme for invalidating old code has been made more accurate and
selective. Moreover, Julia and its standard libraries received a thorough
makeover to help type inference arrive at a concrete answer more often. The
result is a leaner, faster Julia that is far more impervious to method
invalidation, and feels considerably more responsive and nimble in interactive

> sessions. Related blog post: https://julialang.org/blog/2020/08/invalidations/.

Reviewer 2: URLify:
```suggestion
sessions. Related blog post: <https://julialang.org/blog/2020/08/invalidations/>.
```

~~~

For packages that are being developed, given that their code will be changed by
other mechanisms than Pkg, this new workflow won’t automatically avoid
encountering the standard code-load time precompilation. However non-dev-ed
dependencies of those packages will be kept ready to load, so top-level
precompilation at load time should remain lower for dev-ed packages.

## Compile time percentage

_Ian Butterworth_

A small change that should help understanding of one of Julia’s quirks for
newcomers is that the timing macro `@time` and its verbose friend `@timev` now
report if any of the reported time has been spent on compilation.[^1]

```julia-repl
julia> x = rand(10,10);

julia> @time x * x;
  0.540600 seconds (2.35 M allocations: 126.526 MiB, 4.43% gc time, 99.94% compilation time)

julia> @time x * x;
  0.000010 seconds (1 allocation: 896 bytes)
```
Given Julia’s Just In Time (JIT) / Just Ahead Of Time (JAOT) compilation, the
first time code is run the compilation overhead is often substantial, with big
speed improvements seen in subsequent calls. This change highlights that
behavior, serving as both a reminder and a tool for rooting out unwanted

> compilation effort i.e. over-specialized code

Reviewer 3: ```suggestion
compilation effort i.e. over-specialized code
```
seems to be missing punctuation

In 1.6, the scheme for invalidating old code has been made more accurate and
selective. Moreover, Julia and its standard libraries received a thorough
makeover to help type inference arrive at a concrete answer more often. The
result is a leaner, faster Julia that is far more impervious to method
invalidation, and feels considerably more responsive and nimble in interactive
sessions. Related blog post: <https://julialang.org/blog/2020/08/invalidations/>.

## Compiler latency reduction

_Jeff Bezanson_

## Tooling to help optimize packages for latency

_Nathan Daly & Tim Holy_

Julia 1.6, in conjunction with SnoopCompile v2.2.0 or higher, features new tools
for compiler introspection, especially (but not exclusively) for type inference.
Developers can use the new tools to profile type inference and determine how
particular package implementation choices interact with compilation time. Early
adopters have used these tools to eliminate anywhere from a few percent to the
large majority of first-use latency.

- Related blog post: <https://julialang.org/blog/2021/01/precompile_tutorial/>
- Documentation: <https://timholy.github.io/SnoopCompile.jl/stable/>

## Binary loading speedups

> _Elliot Saba & Mosè Giordano_

Reviewer 4: @giordano can you take a pass over this and see if it looks good to you?

reliability of installation came at a cost, however, which was _vastly_
increased load times as compared to the bad old days when Julia packages would
blindly `dlopen()` libraries and load whatever libraries happened to be sitting
on the library search path.  To illustrate the issue, in Julia 1.4, loading the
GTK+3 stack required **8 seconds** when it used to take around **500ms** on the
same machine.  Through many months of hard work and careful investigation, we
are pleased to report that the same stack of libraries now takes less than
**200ms** to load on Julia v1.6.

The cause of this slowdown was multi-faceted and spread across many different
layers of the Julia ecosystem.  Part of the issue was general compiler latency,
which has been a focus of the compiler team for some time now, as evidenced by
Jeff's post in this blogpost.  Another major piece though was general overhead
incurred by having so many small JLL packages providing bindings; there was
significant overhead in the loading of each package.  In particular, there was
code inferrence, code generation and datastructure loading that needed to be

> eliminated if the JLL packages were to be lightweight enough to not effect

Reviewer 3: ```suggestion
eliminated if the JLL packages were to be lightweight enough to not affect
```

Reviewer 4: Although I know the difference, I apparently don't know it well enough for the knowledge to effect me.

reverting a change, committing a new fix, or revising a proposed patch until it
passes. Given this simple policy, it’s difficult to see how a project could be
ambushed by persistent test failures. And yet that’s exactly what happened to
us: over time, we ended up in a state where a high percentage of test runs
failed, usually with just a single obscure test case failing.

Several factors contributed to this predicament. First, the base Julia test
suite is quite large and covers a wide range of functionality, from parsing and
compiling to linear algebra, package management, sockets, handling file system
events, and more. With that much surface area, we were likely to end up with a
handful of rare bugs, or failures due to overly-fragile tests. We run easily
over a hundred builds per day, so even failure rates of 0.1% would appear often
enough to cause problems. Timing-sensitive tests are a classic example, e.g.
testing that a one-second timeout indeed happens after approximately one second.
On hosted VMs in particular, timing can be far more variable than what you would
ever see on dedicated hardware. A one-second timeout can, unfortunately, take

> more than 60 seconds on a heavily loaded VM.

Reviewer 3: This could use a concluding sentence or two to summarize the progress. This does a good job of laying out the problem, and mentions that you put welcome work into it, but it doesn't say anything about how many cases were fixed, or how much of a change there was in the frequency of such problems, etc.

Reviewer 5: Working on it! I put this aside for a couple weeks but I'll add more soon.

Reviewer 3: Your addition is great.

time. In 1.6, this package precompilation is faster and happens before you leave
`pkg>` mode. Until now, precompilation took place solely as a single-processed
sequence, precompiling dependencies one-by-one when needed during the linear
code loading process when a package is `using`/`import`-ed for the first time.

The olden days of <= 1.5:
```julia-repl
(v1.5) pkg> add DifferentialEquations
...
julia> @time using DifferentialEquations
[ Info: Precompiling DifferentialEquations [0c46a032-eb83-5123-abaf-570d42b7fbaa]
  474.288251 seconds …
```
In 1.6, `pkg>` mode gains a heavily parallelized precompile operation that is
auto-invoked after package actions, to keep the active environment ready to
load.
```julia-repl
(v1.6) pkg> add DifferentialEquations
...
Precompiling project...
  Progress [========================================>]  112/112
112 dependencies successfully precompiled in 72 seconds

julia> @time using DifferentialEquations
  4.995477 seconds …
```
Whereas the previous code loading precompilation process took ~8 minutes to
precompile and load without indicating progress while doing so, the new

> mechanism took 1m12s to precompile while showing progress through the

Reviewer: ```suggestion
mechanism took a little over 1 minute to precompile while showing progress through the
```

[`JLLWrappers.jl`](https://github.com/JuliaPackaging/JLLWrappers.jl).  This
package provides macros that auto-generate the bindings necessary for a JLL
package, and do so by using the minimum number of functions and datastructures
possible.  By limiting the number of backedges and datastructures, as well as
centralizing the template pieces of code that each JLL package uses, we are
able to not only vastly improve load times, but improve compile times as well!
As an added bonus, improvements to JLL package APIs can now be made directly in
`JLLWappers.jl` without needing to re-deploy hundreds of JLLs.  Because these
JLL packages only define a thin wrapper around simple, lightweight functions
that load libraries and return paths and such, they do not benefit from the
heavy optimization that most Julia code undergoes.  One final piece of the
optimization puzzle was therefore to disable optimizations and use the new
[per-module optimization levels](https://julialang.org/blog/2020/08/julia-1.5-highlights/#per-module_optimization_levels)
functionality to reduce the amount of time spent generating a very small
amount of code, saving precious seconds.

~~~

> <iframe width="900" height="800" frameborder="0" scrolling="no" src="//plotly.com/~staticfloat/76.embed"></iframe>

Reviewer 2: Maybe something like
```suggestion
<iframe width="900" height="800" frameborder="0" scrolling="no" src="https://chart-studio.plotly.com/~staticfloat/76.embed"></iframe>
```
?

able to not only vastly improve load times, but improve compile times as well!
As an added bonus, improvements to JLL package APIs can now be made directly in
`JLLWappers.jl` without needing to re-deploy hundreds of JLLs.  Because these
JLL packages only define a thin wrapper around simple, lightweight functions
that load libraries and return paths and such, they do not benefit from the
heavy optimization that most Julia code undergoes.  One final piece of the
optimization puzzle was therefore to disable optimizations and use the new
[per-module optimization levels](https://julialang.org/blog/2020/08/julia-1.5-highlights/#per-module_optimization_levels)
functionality to reduce the amount of time spent generating a very small
amount of code, saving precious seconds.

~~~
<iframe width="100%" height="450" frameborder="0" scrolling="no" src="https://chart-studio.plotly.com/~staticfloat/76.embed"></iframe>
~~~

The interplay between compiler improvements and the benefits that `JLLWrappers`
affords were [well-recorded](https://github.com/JuliaGraphics/Gtk.jl/issues/466#issuecomment-716058685)
during the development process, and showcase a speedup of load times for the

> original, non-JLLWrapperized `GTK3_jll` packge from its peak at `6.73` seconds

Reviewer 6: ```suggestion
original, non-JLLWrapperized `GTK3_jll` package from its peak at `6.73` seconds
```

While there are some minor improvements for lazy loading of shared libraries
and such in the pipeline, we are confident that this work will provide a strong
foundation for Julia's binary packaging story for the forseeable future.

## Downloads & NetworkingOptions

_Stefan Karpinski_

In previous releases, when you download something in Julia, either directly,
using the `Base.download` function, or indirectly when using `Pkg`, the actual
downloading was done by some external process—whichever one of `curl`, `wget`,
`fetch` or `PowerShell` happened to be available on your system. The fact that
this frankendownload feature worked at all was something of a miracle, that only
worked due to much fussy command-line-option finessing of over the years. And
while this did mostly work, there were some major drawbacks to this approach.

1. **It’s slow.** Starting a new process for each download is expensive; but
   worse, those processes can’t share TCP connections or reuse already
   negotiated TLS connections, so every download needs to do the while TCP
   SYN/ACK song and dance and then also do the TLS secret handshake, all of

>    which which takes a lot of time.

Reviewer 6: ```suggestion
   which takes a lot of time.
```

2. **It’s inconsistent.** Since the exact way things got downloaded depended on
   what happens to be installed on your system, download behavior was terribly
   inconsistent. Downloads that work on one system might not work on another
   one. Moreover, any issues someone might have, inevitably end up out of scope
   for Julia to fix — the typical answer is "fix your system

>    `curl`/`wget`/whatever," which is not a very satifactory solution for someone

Reviewer 6: ```suggestion
   `curl`/`wget`/whatever," which is not a very satisfactory solution for someone
```

`pkg>` mode. Until now, precompilation took place solely as a single-processed
sequence, precompiling dependencies one-by-one when needed during the linear
code loading process when a package is `using`/`import`-ed for the first time.

The olden days of <= 1.5:
```julia-repl
(v1.5) pkg> add DifferentialEquations
...
julia> @time using DifferentialEquations
[ Info: Precompiling DifferentialEquations [0c46a032-eb83-5123-abaf-570d42b7fbaa]
  474.288251 seconds …
```
In 1.6, `pkg>` mode gains a heavily parallelized precompile operation that is
auto-invoked after package actions, to keep the active environment ready to
load.
```julia-repl
(v1.6) pkg> add DifferentialEquations
...
Precompiling project...
  Progress [========================================>]  112/112
112 dependencies successfully precompiled in 72 seconds

julia> @time using DifferentialEquations
  4.995477 seconds …
```
Whereas the previous code loading precompilation process took ~8 minutes to
precompile and load without indicating progress while doing so, the new
mechanism took a little over 1 minute to precompile while showing progress through the

> dependencies, then the first time the package is loaded it happens at full

Reviewer 7: Maybe my brain is melted at the moment, but the 2nd part of this sentence confuses me @IanButterworth "then the first time the package is loaded it happens at full speed."

What is happening at full speed? Perhaps a stand alone sentence here

Reviewer: ```suggestion
dependencies. Then the first time the package is loaded, it loads at full
```

code gets _invalidated_, marking it as unsuitable for use; second, as needed the
code is again compiled from scratch taking account of the new methods and types.

Earlier versions of Julia were somewhat conservative, and invalidated old code
in some circumstances where there was no actual change in dispatch. Moreover,
there were many places where Julia and its standard libraries were written in a
way that defeated Julia’s type-inference. Because the compiler sometimes had to
invalidate code just because a new method _might_ apply, any uncertainty about
types magnifies the risk and frequency of invalidation. In older versions of
Julia, the combination of these effects made invalidation widespread: just
loading certain packages led to invalidation of up to 10% of Julia’s precompiled
code. The delay for recompilation could sometimes make interactive sessions feel
sluggish. When invalidation occurred in Julia’s package-loading code, it also
delayed loading of the next package, contributing to long waits for `using
SomePkg` when `SomePkg` depends on other packages.

In 1.6, the scheme for invalidating old code has been made more accurate and

> selective. Moreover, Julia and its standard libraries received a thorough

Reviewer 7: ```suggestion
selective. Additionally, Julia and its standard libraries received a thorough
```
@timholy

Reviewer: Doesn't seem necessary

makeover to help type inference arrive at a concrete answer more often. The
result is a leaner, faster Julia that is far more impervious to method
invalidation, and feels considerably more responsive and nimble in interactive
sessions. Related blog post: [Analyzing sources of compiler latency in Julia:
method invalidations](https://julialang.org/blog/2020/08/invalidations/).

## Compiler latency reduction

_Jameson Nash and Jeff Bezanson_

> In addition to making our library code more compiler-friendly, we continue

Reviewer 7: Maybe this is just my lack of understanding, but what is library code vs regular code? Perhaps worth mentioning?

Reviewer 8: All code is library code

to try to speed up the compiler itself. This remains one of our main
technical challenges. In this release there aren't any major
breakthroughs, but we do have some modest improvements due to work on the
method table data structure.

> Method specificity is a partial order, and prior to 1.6 we stored methods in

Reviewer 7: ```suggestion
Method specificity is stored in partial order, prior to 1.6, we stored methods in
```

Reviewer 7: If we keep the same wording here, perhaps adding a link to: https://en.wikipedia.org/wiki/Partially_ordered_set would help anyone scratching their head as I was

them ourselves using the [`BinaryBuilder.jl`](https://github.com/JuliaPackaging/BinaryBuilder.jl)
framework.  Libraries built from `BinaryBuilder.jl` are most often used through
so-called JLL packages which provide a standardized API that Julia packages can
use to access the provided binaries.  This ease of use and reliability of
installation resulted in _vastly_ increased load times as compared to the bad
old days when Julia packages would blindly `dlopen()` libraries and load
whatever libraries happened to be sitting on the library search path.  To
illustrate the issue, in Julia 1.4, loading the GTK+3 stack required **7
seconds** when it used to take around **500ms** on the same machine.  Through
many months of hard work and careful investigation, we are pleased to report
that the same stack of libraries now takes less than **200ms** to load when
using Julia v1.6 on the same machine.

The cause of this slowdown was multi-faceted and spread across many different
layers of the Julia ecosystem.  Part of the issue was general compiler latency,
which has been a focus of the compiler team for some time now, as evidenced by

> Jeff's post in this blogpost.  Another major piece though was general overhead

Reviewer 7: ```suggestion
Jeff's section in this blogpost.  Another major piece though was general overhead
```

incurred by having so many small JLL packages providing bindings; there was
significant overhead in the loading of each package.  In particular, there was

> code inferrence, code generation and datastructure loading that needed to be

Reviewer 7: ```suggestion
code inference, code generation and data structure loading that needed to be
```

Reviewer 7: was going to suggest an Oxford comma but didn't want to start a war

eliminated if the JLL packages were to be lightweight enough to not affect
overall load times.  In our experiments, we found that one of the largest

> sources of package load times was in the deserialization of backedge

Reviewer 7: ```suggestion
sources of package load times was in the deserialization of back edge
```

information, the links from functions in `Base` back to our packages that would
cause our functions to be recompiled if there was an invalidation effecting
that `Base` function.  As counter-intuitive as it may seem, simply using
a large number of functions from `Base` can very quickly balloon the
precompilation cache files for your package, causing an increase in loading
time!  While the increase itself is small, (`3-10ms` at the worst) when you are
loading many dozens of JLL packages, this adds up quickly.

Our work to slim JLL packages down resulted in the creation of a new package,
[`JLLWrappers.jl`](https://github.com/JuliaPackaging/JLLWrappers.jl).  This
package provides macros that auto-generate the bindings necessary for a JLL

> package, and do so by using the minimum number of functions and datastructures

Reviewer 7: ```suggestion
package, and do so by using the minimum number of functions and data structures
```

> possible.  By limiting the number of backedges and datastructures, as well as

Reviewer 7: ```suggestion
possible.  By limiting the number of back edges and data structures, as well as
```

Reviewer: I think it's ok as-is. `backedge` is used elsewhere in blogposts. `julia datastructure` has less hits on google though 🤷‍♂️

Reviewer 9: I would keep `backedge` since I see the core devs use that spelling all the time.

But I would change datastructure to data structure.

Reviewer: Ok. I'll make that change

on Julia v1.4 down to `2.34` seconds on Julia v1.6, purely from compiler
improvements.  If you haven't thanked your local compiler team today, you
probably should.  Using the slimmed-down JLLWrappers implementation of all
relevant JLLWrappers packages results in a further lowering of load time down
to a blistering `140ms`.  End-to-end, this means that this work effected a 
roughly **`50x` speedup** in load times for large trees of binary artifacts.
While there are some minor improvements for lazy loading of shared libraries
and such in the pipeline, we are confident that this work will provide a strong
foundation for Julia's binary packaging story for the forseeable future.

## Downloads & NetworkingOptions

_Stefan Karpinski_

In previous releases, when you download something in Julia, either directly,
using the `Base.download` function, or indirectly when using `Pkg`, the actual
downloading was done by some external process—whichever one of `curl`, `wget`,
`fetch` or `PowerShell` happened to be available on your system. The fact that
this frankendownload feature worked at all was something of a miracle, that only

> worked due to much fussy command-line-option finessing of over the years. And

Reviewer 7: ```suggestion
worked due to much fussy command-line-option finessing over the years. And
```

Reviewer: `of` seems fine. The comma, perhaps. Your call @logankilpatrick

methods to previously-defined functions, and use previously-defined methods on
new types. Sometimes, these new entities force Julia to recompile code to
account for changes in dispatch. This happens in two steps: first, “outdated”
code gets _invalidated_, marking it as unsuitable for use; second, as needed the
code is again compiled from scratch taking account of the new methods and types.

Earlier versions of Julia were somewhat conservative, and invalidated old code
in some circumstances where there was no actual change in dispatch. Moreover,
there were many places where Julia and its standard libraries were written in a
way that defeated Julia’s type-inference. Because the compiler sometimes had to
invalidate code just because a new method _might_ apply, any uncertainty about
types magnifies the risk and frequency of invalidation. In older versions of
Julia, the combination of these effects made invalidation widespread: just
loading certain packages led to invalidation of up to 10% of Julia’s precompiled
code. The delay for recompilation could sometimes make interactive sessions feel
sluggish. When invalidation occurred in Julia’s package-loading code, it also

> delayed loading of the next package, contributing to long waits for using
> SomePkg when SomePkg depends on other packages.

Reviewer 3: ```suggestion
delayed loading of the next package, contributing to long waits for `using
SomePkg` when `SomePkg` depends on other packages.
```

a little progress on it in every release. This time, there aren't any major
breakthroughs, but we do have some modest improvements due to work on the
method table data structure.

Method specificity is a partial order, and prior to 1.6 we stored methods in
sorted order. We also attempted to identify ambiguous methods on insertion,
hoping to avoid repeating the work for each future query.
Unfortunately, sorting a partial order requires quadratic time, and this time
began to show up prominently during package loading (when a package's methods
need to be inserted into the currently-active method tables).

We improved things by making the process lazier, moving sorting and ambiguity
detection into the algorithm for finding matching methods.
This algorithm runs very often, so it was not at first intuitive that this
change would help.
But the key is that the vast majority of queries are for specific enough
types that most possible matches can be eliminated easily, leaving
many fewer inputs to the most expensive steps.

The main visible improvement here is to package loading, which is now often
significantly faster. For example, `using Plots` goes from about 6 seconds to

> about 3 seconds for me (after precompilation).

Reviewer 3: This 2x bundles together the advances due to method sorting with the fact that we don't invalidate `Base.require` or `Pkg` methods so often. See whether you like the partial description in the invalidations section and whether we can combine these usefully (I wrote it anticipating you'd have something like this here, but obviously didn't know exactly how it would come out).

us: over time, we ended up in a state where a high percentage of test runs
failed, usually with just a single obscure test case failing.

Several factors contributed to this predicament. First, the base Julia test
suite is quite large and covers a wide range of functionality, from parsing and
compiling to linear algebra, package management, sockets, threads, handling file system
events, and more. With that much surface area, we were likely to end up with a
handful of rare bugs, or failures due to overly-fragile tests. We run easily
over a hundred builds per day, so even failures with a rate of 0.1% would appear often
enough to be disruptive. Timing-sensitive tests are a classic example, e.g.
testing that a one-second timeout indeed happens after approximately one second.
On hosted VMs in particular, timing can be far more variable than what you would
ever see on dedicated hardware. A one-second timeout can, unfortunately, take
more than 60 seconds on a heavily loaded VM.

After much debugging, including infrastructure work to run tests under
[rr](https://julialang.org/blog/2020/05/rr) by default, we were able to

> identify and fix a significant number of issues. Here is a representative sample:

Reviewer 3: ```suggestion
identify and fix many issues. Here is a representative sample:
```
(Statisticians have special reverence for "significant" and we have a lot of statisticians reading this.)

other mechanisms than Pkg, this new workflow won’t automatically avoid
encountering the standard code-load time precompilation. However non-dev-ed
dependencies of those packages will be kept ready to load, so top-level
precompilation at load time should remain lower for dev-ed packages.

## Compile time percentage

_Ian Butterworth_

A small change that should help understanding of one of Julia’s quirks for
newcomers is that the timing macro `@time` and its verbose friend `@timev` now
report if any of the reported time has been spent on compilation.[^1]

```julia-repl
julia> x = rand(10,10);

julia> @time x * x;
  0.540600 seconds (2.35 M allocations: 126.526 MiB, 4.43% gc time, 99.94% compilation time)

julia> @time x * x;
  0.000010 seconds (1 allocation: 896 bytes)
```
Given Julia’s Just In Time (JIT) / Just Ahead Of Time (JAOT) compilation, the
first time code is run the compilation overhead is often substantial, with big
speed improvements seen in subsequent calls. This change highlights that
behavior, serving as both a reminder and a tool for rooting out unwanted
compilation effort i.e. over-specialized code.

> [^1]: Note that in some cases the system will look inside the `@time` expression
> and compile some of the called code before execution of the top-level expression begins.
> When that happens, some compilation time will not be counted.

Reviewer: ```suggestion
[^1]: Note that in some cases the system will look inside the `@time` expression and compile some of the called code before execution of the top-level expression begins. When that happens, some compilation time will not be counted. To include this time you can run `@time @eval ...`
```
Two changes: 
- Has to be a single line to be captured in the tip box
- Added tip to try `@time @eval ...` as per https://github.com/JuliaLang/julia/pull/39802

Load times and compilation pauses continue to be a significant nuissance for
Julia users, and one of our main technical challenges. We try to make at least
a little progress on it in every release. This time, there aren't any major
breakthroughs, but we do have some modest improvements due to work on the
method table data structure.

Method specificity is a partial order, and prior to 1.6 we stored methods in
sorted order. We also attempted to identify ambiguous methods on insertion,
hoping to avoid repeating the work for each future query.
Unfortunately, sorting a partial order requires quadratic time, and this time
began to show up prominently during package loading (when a package's methods
need to be inserted into the currently-active method tables).

We improved things by making the process lazier, moving sorting and ambiguity
detection into the algorithm for finding matching methods.
This algorithm runs very often, so it was not at first intuitive that this
change would help.
But the key is that the vast majority of queries are for specific enough
types that most possible matches can be eliminated easily, leaving
many fewer inputs to the most expensive steps.

> 

Reviewer 10: Should https://github.com/JuliaLang/julia/pull/36208 and `Base.Experimental.@compiler_options` be mentioned here as well?
